### PR TITLE
BIP-110: advance to Complete status

### PR DESCRIPTION
**Pull request by dathonohm**
Also emphasize UTXO grandfathering and restore OP_IF byte-saving note.

### CODE REVIEW

**Review by jonatack — COMMENTED** [review]
Test vectors missing or planned to be added? Per BIP 3:

"Specification BIPs must come with or refer to a working reference implementation and comprehensive test vectors before they can be moved to Complete. Subsequently, the BIP’s content should only be adjusted in minor details, e.g., to improve language, clarify ambiguities, backfill omissions in the specification, add test vectors for edge cases, or address other issues discovered as the BIP is being adopted."

Edit: I suppose that updating the tests to pass in the reference implementation could fulfill this, but has test coverage specific to the changes been added?

Consider also if you think any non-minor changes to the Specification may still be needed before advancing the status to complete.

Finally, consider if you want to add a changelog. BIP 3 states that "Version 1.0.0 is used to label the promotion to Complete. A Changelog section may be introduced during the Draft phase to record significant changes (using versions 0.x.y)."

File: bip-0110.mediawiki [hunk 1]
```diff
@@ -30,7 +30,11 @@ Blocks during a temporary, one-year deployment are checked with these additional
 # Tapscripts including OP_SUCCESS* opcodes anywhere (even unexecuted) are invalid.
 # Tapscripts executing the OP_IF or OP_NOTIF instruction (regardless of result) are invalid.
 
-Inputs spending UTXOs that were created before the activation height are exempt from the new rules.
+===UTXO grandfathering===
+
+'''Inputs spending UTXOs that were created before the activation height are exempt from all of the new rules.'''
+This grandfathering ensures that no existing coins can be frozen or rendered unspendable by this softfork: any UTXO confirmed before activation can always be spent exactly as it could before, throughout the entire deployment.
+The new rules apply only to UTXOs created at or after the activation height.
```
Comment (jonatack) [inline bip-0110.mediawiki:37]: Do you wish to wait for review feedback on this change to the Specification, before updating to Complete?

File: bip-0110.mediawiki [hunk 2]
```diff
@@ -177,7 +181,7 @@ Yes:
...
 # Upgrade hooks are not available for other softforks. As softforks adding new opcodes typically need at least a year to activate, this shouldn't be a practical issue.
-# Some wallet software such as Miniscript habitually creates Tapleaves containing OP_IF. To mitigate the risk of these funds being frozen for a year, this proposal exempts inputs that spend outputs that were created before activation, and provides a two-week grace period between lock-in and activation, to give users time to prepare.
+# Despite there being no known use case that requires it, some wallet software such as Miniscript habitually creates Tapleaves containing OP_IF, which in some hypothetical cases can require less data than the equivalent construction using additional Tapscripts. If such a use case is discovered, this simply increases the fee for that construction until the softfork expires.
```
Comment (lifofifoX) [inline bip-0110.mediawiki:184]: On-chain transparency of spending conditions is a well known use case.

File: bip-0110.mediawiki (hunk 1 above)
Comment (dathonohm) [reply bip-0110.mediawiki:37]: It's not a change, just a clarification. Many commenters were apparently unaware of UTXO grandfathering, so I decided to make it more prominent.

File: bip-0110.mediawiki (hunk 2 above)
Comment (dathonohm) [reply bip-0110.mediawiki:184]: It is theoretical, and there are no known non-experimental implementations, to my knowledge.

Besides, P2WSH already allows on-chain transparency of spending conditions, for fewer bytes.

**Comment by dathonohm** [comment]
Hi @jonatack:

>Edit: I suppose that updating the tests to pass in the reference implementation could fulfill this, but has test coverage specific to the changes been added?

Yes, there is a test file in the reference implementation that specifically targets the spec: https://github.com/dathonohm/bitcoin/blob/uasf-modified-bip9/test/functional/feature_rdts.py. I can update the BIP document to point there, if you like.

All unit and functional tests have been passing since December.

>Consider also if you think any non-minor changes to the Specification may still be needed before advancing the status to complete.

The spec has been public and under active review since October, and stable since ~January. A seventh of the network is already running this spec and mandatory signaling begins in a month and a half; changing the spec now would be highly disruptive, so I don't think it's a good idea, unless there is some kind of severe flaw discovered (unlikely because of how simple the changes are).

>Finally, consider if you want to add a changelog. BIP 3 states that "Version 1.0.0 is used to label the promotion to Complete. A Changelog section may be introduced during the Draft phase to record significant changes (using versions 0.x.y)."

Does this include past changes, or just changes made from this point moving forward?

**Comment by jonatack** [comment]
> Yes, there is a test file in the reference implementation that specifically targets the spec: [dathonohm/bitcoin@uasf-modified-bip9/test/functional/feature_rdts.py](https://github.com/dathonohm/bitcoin/blob/uasf-modified-bip9/test/functional/feature_rdts.py). I can update the BIP document to point there, if you like.

Good idea.

> Does this include past changes, or just changes made from this point moving forward?

Up to you, I think.

File: bip-0110.mediawiki (hunk 2 above)
Comment (lifofifoX) [reply bip-0110.mediawiki:184]: How is it theoretical? It's literally how it works. IMO, you should try to use Bitcoin a bit more often under real life constraints.

Comment (dathonohm) [reply bip-0110.mediawiki:184]: The fact that a use case is *possible* does not mean it is practical. All use cases invalidated by BIP-110 are accordingly pathological.

**Comment by dathonohm** [comment]
@jonatack - I've pushed the Changelog and Test vectors sections.

I've also generated some [test vectors and a script that verifies them against a node](https://github.com/dathonohm/bips/commit/83a3af65ff01a12d186e5943da1032a2bcba17eb) (as well as instructions to the reviewer for how to run the script). I have verified the vectors against the reference implementation ([merged into Knots](https://github.com/bitcoinknots/bitcoin/pull/238) a few weeks ago) and also the [port to Core](https://github.com/bip110-core/bitcoin/pull/2) (not merged).


I wasn't sure if it would be useful to have the BIP-110 spec rules in JSON format so I put them on a separate branch. I can push those changes here if you think it's worth it.

**Comment by jonatack** [comment]
The test vectors and script wouldn't hurt. This PR appears otherwise fine as-is, or you can add them first before advancing here to Complete. Let me know.

**Comment by dathonohm** [comment]
@jonatack - pushed.

**Comment by jonatack** [comment]
Thanks for updating. The new additions look well-documented.

File: bip-0110.mediawiki
Comment (jonatack) [inline bip-0110.mediawiki:10]: I overlooked to mention to please add a `Version: 1.0.0` header on the line after Discussion.

BIP 3: "After a BIP receives a Changelog, the Preamble must indicate the latest version in the Version header."

Comment (jonatack) [reply bip-0110.mediawiki:10]: (We should add a CI check for missing Version header when a Changelog is present in `scripts/buildtable.pl`)

Comment (jonatack) [reply bip-0110.mediawiki:10]: Doing these as a follow-up.

Comment (jonatack) [reply bip-0110.mediawiki:10]: Added a Version header in #2203 (edit: and further details to the changelog entry)

Comment (dathonohm) [reply bip-0110.mediawiki:10]: Thank you!
